예외 처리 초기 구성

jieun-dev1 opened the pull request:
### **개요**

모든 예외는 GlobalControllerAdvice 에서 처리 됩니다. 
초기 구성으로 User 패키지 전용 예외를 생성하고 서비스/컨트롤러에 적용했습니다. 

### **이슈 링크**

#5 

### **작업 사항 (변경 사항)**
GlobalControllerAdvice, ErrorController, BaseException, UnExpectedException, ErrorResult 및 User 패키지 전용 Exception추가. 


### 테스트 내역 (o/x)
포스트맨 테스트
예상치 못한 예외는 Internal Server Error, 사전에 정의한 에러는 해당 에러가 요청한 JSON 형식에 맞게 응답(o)

-register 회원 가입 로직에서 중복인 경우 DuplicatedUserException 발생
-없는 loginId를 조회할 경우 UserNotFound 리턴 
-에러 수준 별 로그가 찍히는 것도 확인. 

### **리뷰 받고 싶은 내용**

구체적으로 조언을 구하고 싶은 내용은 없으나, 개선할 점이 있다면 피드백 부탁드립니다.

jieun-dev1 committed:
feat(common): BaseException, UnExpectedException, ErrorResult 생성

사용자 지정 예외 BaseException
예상치 못한 예외를 받아줄 UnExpectedException
반환 에러 형식인 Error Result 생성

jieun-dev1 committed:
feat(common): ErrorController 생성

아직 Service 로 구현되지 않았지만, 추후 쓰일 에러, unexpectedException 테스트하기 위함

jieun-dev1 committed:
feat(user): User 패키지에서 쓰일 Exception 생성

BadRequest: 아직 서비스에서 정의되지 않았지만 잘못된 요청 시
UserNotFoundException: 유저 검색 시 유저가 없을 때
DuplicatedUserException: 회원 가입 시도 시 중복 유저일때

jieun-dev1 committed:
feat(common): 공통 Controller Advice 지정

에러 별로 처리하기 위해 핸들러 매핑
예외가 발생했을 떄, 지정한 ErrorResult 형태로 반환
에러의 수준 별 로그 처리

jieun-dev1 committed:
feat(user): 정의한 user패키지 exception 을 controller 및 서비스 로직에 반영

sgo8308 commented on src/main/java/com/marketgola/common/exception/GlobalControllerAdvice.java:
```diff
@@ -0,0 +1,68 @@
+        LogLevel level = e.getLogLevel();
+        String requestURI = request.getRequestURI();
+        String method = request.getMethod();
+        String logMessage = String.format("[EXCEPTION LOGGING] errorCode:{} method: {} path:{}");
```
String.format()으로 감싸는 것과 감싸지 않는 것은 무슨 차이가 있을까요?

derek-Flab replied:
아 여기서는 String.format을 명시해주실 필요는 없을 것 같네요!

jieun-dev1 replied:
알고보니 이미 log.error 에서 인자와 함께 String을 만들어주고 있었어서, String.format 을 없애도 결과가 같더라고요. :)반영했습니다.

sgo8308 commented on src/main/java/com/marketgola/common/exception/GlobalControllerAdvice.java:
```diff
@@ -0,0 +1,68 @@
+        } else if (level == LogLevel.INFO) {
+            log.info(logMessage, e.getStatus(), method, requestURI);
+        } else if (level == LogLevel.DEBUG) {
+            log.debug(logMessage, e.getStatus(), method, requestURI);
```
(logMessage, e.getStatus(), method, requestURI); 이 부분이 4개나 중복되는데 어떻게하면 없앨 수 있을까요?

derek-Flab replied:
메시지는 위에서 사용해보신 String.format으로 처리해서 중복 코드를 줄여볼 수 있지 않을까요?

derek-Flab replied:
`log.xxx("메시지", 변수1, 변수2, e)`
뒤에 e를 꼭 던져주세요. 스택 트레이스가 로그에 찍히도록요.

jieun-dev1 replied:
String.format 반영해보았습니다. 더 개선할 아이디어 있으시면 공유부탁드려요! 
exception 추가해서 스택 트레이스 찍히는 것 확인했습니다.

sgo8308 commented on src/main/java/com/marketgola/user/exception/BadRequestException.java:
```diff
@@ -0,0 +1,17 @@
+public class BadRequestException extends BaseException {
+
+    public BadRequestException(HttpStatus status, LogLevel logLevel, String message) {
+        super(HttpStatus.BAD_REQUEST, LogLevel.ERROR, message);
```
이렇게 코드를 짤 경우 status와 logLevel을 파라미터로 받을 필요가 있을까요?

derek-Flab replied:
네 지우님 말씀처럼 불필요한 인자를 받으신거 같고, 다른 인자 조합의 생성자를 만드셔서 제공하도록 하면 좋을 것 같습니다.

jieun-dev1 replied:
다시보니  status와 logLevel이 인자가 사용되지 않고 있었더라고요! message 만으로도 충분한 것 같아서 수정하였습니다~

sgo8308 commented on src/main/java/com/marketgola/user/exception/UserNotFoundException.java:
```diff
@@ -0,0 +1,16 @@
+public class UserNotFoundException extends BaseException {
+
+    public UserNotFoundException(String message) {
+        super(message, HttpStatus.NOT_FOUND, LogLevel.DEBUG);
```
같은 예외가 상황에 따라 DEBUG일 수도 있고 더 높은 수준 일수도 있기 때문에 처음부터 이 예외는 DEBUG level로 하겠어! 라고 정하는 것보다는 예외를 생성할 때 정할 수 있는 게 나아 보입니다.

관련해서 멘토님 리뷰 첨부합니다.
UserNotFoundException이 있다고 예를 들어보죠.
이 유저에게 배송을 하려고하는데 유저 정보가 없어서(예를들어 탈퇴 해버림) 발생한 예외는 fatal exception이 맞는 것 같습니다.
근데, 유저가 단순히 댓글을 적고 탈퇴한 후 댓글을 쓴 유저를 접근하려고 할 때 발생하는 경우는 fatal은 아닐 수 있을 것 같습니다.

이런 경우 같은 예외를 로직에 따라 던질때 로그레벨을 정하고 싶거나, 그런 니즈가 많을 것 같은데요~
이는 결국 생성할때 로그레벨을 지정할 수 있게 하면 좋겠다는 말이죠~

derek-Flab replied:
네네 덧붙여 디폴트로 LogLevel.DEBUG를 지정하는 생성자도 만들고, 
LogLevel을 인자로 받는 생성자도 만들어서 제공하면 예외를 던질 때 유용하게 사용할 수 있을 것 같아요.

sgo8308 reviewed (commented):
화이팅입니다 ㅎㅎ

derek-Flab commented on src/main/java/com/marketgola/common/exception/ErrorController.java:
```diff
@@ -0,0 +1,29 @@
+@RestController
+public class ErrorController {
+
+    @GetMapping("/wronginput")
```
API Path가 임시 인 것으로 알고 있지만, 
uri를 지정할 때 이런 문서 한번 읽어보면 좋을 것 같아요. 
https://restfulapi.net/resource-naming/

(cc. @sgo8308)

derek-Flab commented on src/main/java/com/marketgola/common/exception/ErrorResult.java:
```diff
@@ -0,0 +1,22 @@
+    private final int status; //ex. 404
+    private final String error; //ex. "Not Found",
+    private final String message; //ex. "Data Not Found"
+    private final String url;
```
spring boot에서 기본적으로 제공해주는 error controller 의 예시를 보면 
path라는 용어를 사용하고 있습니다. 예시를 적어보면
```
{
  "timestamp": "2019-02-15T21:48:44.447+0000",
  "status": 404,
  "error": "Not Found",
  "message": "No message available",
  "path": "/123"
}
```

요렇게 동일하게 가도 좋을 것 같아요.

derek-Flab commented on src/main/java/com/marketgola/user/exception/DuplicatedUserException.java:
```diff
@@ -0,0 +1,15 @@
+ */
+public class DuplicatedUserException extends BaseException {
+
+    public DuplicatedUserException(String message) {
```
네이밍에 일관성이 부여되면 좋을 것 같습니다!
User + NotFound조합
User + Duplicdated조합. 
이런식으로요. 네이밍 컨벤션과 일관성을 지켜보면 좋을 것 같습니다!

derek-Flab commented:
오 서로서로 리뷰해주시고 좋네요! 
서로 코드 리뷰를 통해 많은 것을 얻어가실 수 있을 것 같아요.!! 화이팅입니다.

jieun-dev1 committed:
refactor (common, user):Exception 관련 로직 수정

BaseException
-중복 생성된 생성자 제거
-사용되지 않는 생성자 제거

ErrorController
-UserBadRequestException의 HttpStatus 는 경우에 따라 다르게 지정할 필요는 없다고 생각되어, Exception 클래스에서 지정
- 필요에 따라 logLevel 수준을 설정할 수 있도록 생성자 추가

GlobalControllerAdvice
-url 을 공식문서 명칭인 path 로 변경
-String.format 을 사용하여, 중복되는 변수 3개를 줄임
-스택 트레이스가 로그에 찍히도록 log 메서드에 e를 추가

ErrorResult
-dateTime을 공식 문서의 명칭인 timestamp로 변경
-url 을 공식문서 명칭인 path 로 변경

UserService
-예외 클래스의 네이밍 일관성을 위해 UserDuplicatedException으로 수정

jieun-dev1 committed:
refactor (common, user): 지정 예외 처리인 Exception 클래스 수정

공통
-기본 logLevel을 DEFAULT로 수정
-필요에 따라 logLevel 수준을 설정할 수 있도록 생성자 추가.
-네이밍 일관성 위해 BadRequestException 앞에 user 추가
